### PR TITLE
Sort product table (in email, invoices and order view) by reference id

### LINKED ISSUE #17903: Sort product table (in email, invoices and order view) by reference id
Hello, 

for working better and faster is it would be advisable that both in the orders in the emails and in the pdf the products were ordered by reference

so as to speed up the pickup from the warehouse, as well as avoid errors.

_EDIT:_

**Specs:**
Sort the table of products in alphabetical order of references:
- the invoice file, 
- the credit slip file, 
- the delivery slip file, 
- the order preview on the order list, 
- the products table on the order page, 
- the cart summary on the shopping cart page.

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Sort product table (in email, invoices and order view) by reference id
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #17903
| How to test?      | Cf #17903

### CODE REVIEW

**Review by MeKeyCool — COMMENTED** [review]
Hi @Progi1984 .

I think it should be interesting to add a public static function to `PrestaShop\PrestaShop\Core\Domain\ValueObject\QuerySorting` 
```php
    /**
     * @param string $firstSortingCriteria (TODO : description)
     * @param string $secondSortingCriteria (TODO : description)
     * @param string $ascedingOrder (TODO : description)
     * 
     * @return callable function used to sort
     * 
     * @description : This function is used to sort arrays.
     */
    public static function buildHumanSortingFunction(
        string $firstSortingCriteria,
        string $secondSortingCriteria = null,
        string $ascendingOrder = QuerySorting::ASC
    ): callable {

        return function ($a, $b) use ( $firstSortingCriteria, $secondSortingCriteria, $ascendingOrder) {
            $cmp = strnatcmp($a[$firstSortingCriteria], $b[$firstSortingCriteria]);

            if ( ($cmp === 0) && !empty($secondSortingCriteria) ) {
                $cmp = strnatcmp($a[$secondSortingCriteria], $b[$secondSortingCriteria]);
            }

    	    $cmp = (QuerySorting::DESC === $ascendingOrder) ? $cmp : $cmp * -1;
    		
            return $cmp; 
        };
    }
```

This way, we can test and avoid code duplication (with all maintenance troubles it induces).
Usage would be something like this :

```php
    $toSortArray = [
    	[
    		'a' => "A",
    		'b' => "X",
		],[
    		'a' => "a",
    		'b' => "X",
		],
    	[
    		'a' => "À",
    		'b' => "X",
		],
    	[
    		'a' => "a",
    		'b' => "Y",
		],
    	[
    		'a' => "à",
    		'b' => "Y",
		],
    	[
    		'a' => "B",
    		'b' => "X",
		],
    	[
    		'a' => "B",
    		'b' => "Y",
		],
    	[
    		'a' => "b",
    		'b' => "X",
		],
	];
    
    
usort($toSortArray, QuerySorting::buildHumanSortingFunction('a', 'b', QuerySorting::DESC));

var_dump($toSortArray);
``` 

What do you think @Progi1984 ? :)

**Comment by Progi1984** [comment]
@MeKeyCool Fixed with a sorter class.

File: src/Core/Util/Sorter.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+class Sorter
+{
+    public const ORDER_ASC = 'ASC';
+    public const ORDER_DESC = 'DESC';
+
+    /**
+     * @param array<array<string, mixed>> $array
+     * @param string $order
+     * @param string ...$criterias
```
Comment (MeKeyCool) [inline src/Core/Util/Sorter.php:39]: Good idea to let multiple criteria :+1: 

Your naming conventions are good as well :smiley: 

Why don't you make this function static ? As it is perfectly stateless, what is the interest to require a `new` call with all background memory management stuff ?

I wasn't sure with my suggestion about `QuerySorting` class. Is it deprecated ? Maybe I didn't understand its usage ? Maybe we should add some deprecation notice ?

File: tests/Unit/Core/Util/SorterTest.php [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Unit\Core\Util;
+
+use PHPUnit\Framework\TestCase;
+use PrestaShop\PrestaShop\Core\Util\Sorter;
+
+class SorterTest extends TestCase
```
Comment (MeKeyCool) [inline tests/Unit/Core/Util/SorterTest.php:34]: Ilike this :smile:

File: tests/Unit/Core/Util/SorterTest.php [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    public function dataProviderNatural(): iterable
+    {
+        yield [
+            [],
+            [],
+            Sorter::ORDER_ASC,
+        ];
+
+        // Array with one criteria
```
Comment (MeKeyCool) [inline tests/Unit/Core/Util/SorterTest.php:69]: I think I prefer having a dedicated test function to make problem solving easier.
But it's only a question of taste, not blocking for me ^^

File: tests/Unit/Core/Util/SorterTest.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            ],
+            [
+                ['keyA' => 'a', 'keyB' => 'a'],
+                ['keyA' => 'b', 'keyB' => 'b'],
+            ],
+            Sorter::ORDER_ASC,
+            'keyA',
+        ];
+
+        // Array with one criteria which doesn't exist
```
Comment (MeKeyCool) [inline tests/Unit/Core/Util/SorterTest.php:96]: I think this case may log a WARNING. But if I remember well, there is no logging system in Prestashop :thinking:

**Review by MeKeyCool — APPROVED** [review]
@Progi1984 thanks for your work ^^

Few comments but not blocking.

File: src/Core/Util/Sorter.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+     *
+     * @return array
+     */
+    public function natural(array $array, string $order, string ...$criterias): array
+    {
+        usort($array, function ($a, $b) use ($order, $criterias) {
```
Comment (matks) [inline src/Core/Util/Sorter.php:45]: Isn't it a little close to `natsort` ? 😄

**Review by matks — APPROVED** [review]

File: src/Core/Util/Sorter.php (hunk 5 above)
Comment (Progi1984) [reply src/Core/Util/Sorter.php:45]: But `natsort` only sort array of scalar values, not array of array.

**Review by ghost — APPROVED** [review]

**Review by sLorenzini — APPROVED** [review]
hello @Progi1984,

Thank you for this PR.
Tested on develop branch with php 7.3

Products are well sorted by reference id in order details page / invoices pdf and emails:

https://user-images.githubusercontent.com/2854834/182184190-a3b74462-bc25-4357-ad7d-b081cc212387.mp4

QA approved ✅ 

Thanks

**Comment by matthieu-rolland** [comment]
thanx @Progi1984 and @sLorenzini 👍

**Pull request merged**